File: src/dockerode-runner.ts
import Dockerode from "dockerode";
import * as path from "path";
import { createDockerConfig } from "./docker-config.js";

const docker = new Dockerode();

/**
 * Render R Markdown using Dockerode
 * 
 * @param rmdDir Directory containing R Markdown files
 * @param filename Filename of the R Markdown file to render
 * @param outputFormat Output format (html, pdf, or word)
 * @returns Output filename
 */
export async function renderWithDockerode(
  rmdDir: string,
  filename: string,
  outputFormat: "html" | "pdf" | "word" = "html"
): Promise<string> {
  // Get Docker configuration
  const config = createDockerConfig(rmdDir, filename, outputFormat);
  
  try {
    console.error(`Creating Docker container with bind: ${config.binds[0]}`);
    
    // Create a container using the configuration
    const container = await docker.createContainer({
      Image: config.image,
      Cmd: config.command,
      HostConfig: {
        Binds: config.binds
      }
    });

    // Start the container
    await container.start();

    // Wait for the container to finish
    await container.wait();

    // Get the logs
    const logs = await container.logs({
      stdout: true,
      stderr: true
    });

    // Remove the container
    await container.remove();

    // Convert logs to string
    const logsString = logs.toString();
    
    // Extract the output file path
    const outputFileMatch = logsString.match(/Rendered file: ([^\n]+)/);
    if (!outputFileMatch) {
      throw new Error("Failed to extract output file path from logs");
    }

    const outputFilePath = outputFileMatch[1];
    const outputFileName = path.basename(outputFilePath);
    
    return outputFileName;
  } catch (error: any) {
    console.error("Error rendering with Dockerode:", error);
    throw new Error(`Failed to render R Markdown file with Dockerode: ${error.message || String(error)}`);
  }
}

File: dockerode_runner.go
package main

import (
	"context"
	"fmt"
	"io"
	"path/filepath"
	"regexp"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
)

// RenderWithDockerode renders R Markdown using the Docker API
func RenderWithDockerode(rmdDir, filename, outputFormat string) (string, error) {
	// Get Docker configuration
	config := CreateDockerConfig(rmdDir, filename, outputFormat)

	// Create Docker client
	ctx := context.Background()
	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return "", fmt.Errorf("failed to create Docker client: %w", err)
	}
	defer cli.Close()

	fmt.Printf("Creating Docker container with bind: %s\n", config.Binds[0])

	// Create container
	resp, err := cli.ContainerCreate(
		ctx,
		&container.Config{
			Image: config.Image,
			Cmd:   config.Command,
		},
		&container.HostConfig{
			Binds: config.Binds,
		},
		nil,
		nil,
		"",
	)
	if err != nil {
		return "", fmt.Errorf("failed to create container: %w", err)
	}

	// Start container
	if err := cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
		return "", fmt.Errorf("failed to start container: %w", err)
	}

	// Wait for container to finish
	statusCh, errCh := cli.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
	select {
	case err := <-errCh:
		if err != nil {
			return "", fmt.Errorf("error waiting for container: %w", err)
		}
	case <-statusCh:
	}

	// Get logs
	out, err := cli.ContainerLogs(ctx, resp.ID, types.ContainerLogsOptions{
		ShowStdout: true,
		ShowStderr: true,
	})
	if err != nil {
		return "", fmt.Errorf("failed to get container logs: %w", err)
	}
	defer out.Close()

	// Read logs
	logs, err := io.ReadAll(out)
	if err != nil {
		return "", fmt.Errorf("failed to read container logs: %w", err)
	}

	// Remove container
	if err := cli.ContainerRemove(ctx, resp.ID, types.ContainerRemoveOptions{}); err != nil {
		fmt.Printf("Warning: failed to remove container: %v\n", err)
	}

	// Extract output file path from logs
	logsStr := string(logs)
	re := regexp.MustCompile(`Rendered file: ([^\n]+)`)
	matches := re.FindStringSubmatch(logsStr)
	if len(matches) < 2 {
		return "", fmt.Errorf("failed to extract output file path from logs")
	}

	outputFilePath := matches[1]
	outputFileName := filepath.Base(outputFilePath)

	return outputFileName, nil
}
